pkg/server/plugin/notifier/k8sbundle/k8sbundle_watcher.go

```diff
@@ -0,0 +1,149 @@
+	if b.cancel != nil {
+		b.mu.Unlock()
+		return errors.New("already started")
+	}
+	b.cancel = cancel
+	b.mu.Unlock()
+
+	// allow for another start after this function returns
+	defer func() {
+		b.mu.Lock()
+		b.cancel = nil
+		b.mu.Unlock()
+	}()
+
+	config, err := b.p.getConfig()
+	if err != nil {
+		return err
+	}
+	clients, err := b.p.hooks.newKubeClient(config)
+	if err != nil {
+		return err
+	}
+	watchers, validWatcherPresent, err := newWatchers(ctx, config, clients)
+	if err != nil {
+		return err
+	}
+	if !validWatcherPresent {
+		return nil
+	}
+
+	err = b.hooks.watcherEvents(ctx, config, clients, watchers)
+	switch err {
+	case nil, context.Canceled:
+		return nil
+	default:
+		return err
+	}
+}
+
+func (b *bundleWatcher) Stop() {
+	b.mu.Lock()
+	if b.cancel != nil {
+		b.cancel()
+	}
+	b.mu.Unlock()
+}
+
+// watcherFunc watches for new objects that are created with the proper selector and updates the CA Bundle
+func (b *bundleWatcher) watcherEvents(ctx context.Context, c *pluginConfig, clients []kubeClient, watchers []watch.Interface) (err error) {
+	selectCase := newSelectCase(ctx, watchers)
+	for {
+		chosen, recv, _ := reflect.Select(selectCase)
+		if chosen < len(clients) {
+			if err = b.watcherEvent(ctx, c, clients[chosen], recv.Interface().(watch.Event)); err != nil {
+				return k8sErr.New("handling watch event: %v", err)
+			}
+		} else {
+			return ctx.Err()
+		}
+	}
+}
+
+// watchEvent triggers the read-modify-write for a newly created webhook
+func (b *bundleWatcher) watcherEvent(ctx context.Context, c *pluginConfig, client kubeClient, event watch.Event) (err error) {
+	if event.Type == watch.Added {
+		webhookMeta, err := meta.Accessor(event.Object)
+		if err != nil {
+			return err
+		}
+
+		b.p.log.Debug("Setting bundle for new object", "name", webhookMeta.GetName())
+		if err = b.p.updateBundle(ctx, c, client, webhookMeta.GetNamespace(), webhookMeta.GetName()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func newSelectCase(ctx context.Context, watchers []watch.Interface) []reflect.SelectCase {
+	selectCase := []reflect.SelectCase{}
+	for _, watcher := range watchers {
+		if watcher != nil {
+			selectCase = append(selectCase, reflect.SelectCase{
+				Dir:  reflect.SelectRecv,
+				Chan: reflect.ValueOf(watcher.ResultChan()),
+			})
+		} else {
+			selectCase = append(selectCase, reflect.SelectCase{
+				Dir: reflect.SelectRecv,
+			})
+		}
+	}
+	selectCase = append(selectCase, reflect.SelectCase{
+		Dir:  reflect.SelectRecv,
+		Chan: reflect.ValueOf(ctx.Done()),
+	})
+	return selectCase
+}
+
+func newWatchers(ctx context.Context, c *pluginConfig, clients []kubeClient) ([]watch.Interface, bool, error) {
+	watchers := []watch.Interface{}
+	validWatcherPresent := false
+	for _, client := range clients {
+		watcher, err := client.Watch(ctx, c.Label)
+		if err != nil {
+			return nil, false, err
+		}
+		if watcher != nil {
+			validWatcherPresent = true
+		}
+		watchers = append(watchers, watcher)
```

Comment:
Seems a little funny and potentially dangerous to add a `nil` watcher to the list. Why not just only append for non-nil watchers? Then the `validWatcherPresent` bool can go away and a caller knows if there is valid watches simply by checking if `len(watchers) > 0`...

Reply:
I removed the validWatcherPresent bool. [reflect.SelectCase](https://golang.org/pkg/reflect/#SelectCase) allows a chan to be nil, `If Chan is a zero Value, then the case is ignored...`.  In any case, I updated so the watcher is not even called if the `label` config is not set.

Reply:
Even so, the presence of a `nil` watcher in the slice is fragile. Future maintenance could do something with that slice that would cause a panic if it ran into a nil watcher.